src/Agent.Worker/TestResults/ResultsCommandExtension.cs

```diff
@@ -165,14 +184,14 @@ private void ProcessPublishTestResultsCommand(IExecutionContext context, Diction
                     {
                         _executionContext.Warning(StringUtil.Loc("InvalidResultFiles", resultFile, resultReader));
                     }
-                }                
+                }
 
-                //publish run if there are results.
-                if (runResults.Count > 0)
+                finishedParsing.Set();
+
+                if(resultsExist.IsSet)
                 {
-                    string runName = string.IsNullOrWhiteSpace(_runTitle)
-                    ? StringUtil.Format("{0}_TestResults_{1}", _testRunner, buildId)
-                    : _runTitle;
+                    // Waiting for the publishing task to complete.
+                    publishingTask.Wait(cancellationToken);
```

Comment:
`var publishingTask = Task.Run(() => PublishResults(publisher, cancellationToken));`
you already feed in `cancellationToken` to `PublishResults()`, why the `PublishResults()` doesn't handle the cancellation?
i would assume `await publishingTask;`